nixos/modules/system/boot/loader/grub/install-grub.pl

```diff
@@ -572,24 +586,26 @@ sub readGrubState {
     chomp($devices);
     my $efiMountPoint = <FILE>;
     chomp($efiMountPoint);
+    my $extraGrubInstallArgs = <FILE>;
+    # For historical reasons we do not check the values above for un-definedness
+    # (that is, when the state file has too few lines and EOF is reached),
+    # because the above come from the first version of this logic and are thus
+    # guaranteed to be present.
+    $extraGrubInstallArgs = defined $extraGrubInstallArgs ? $extraGrubInstallArgs : '[]'; # empty JSON array
+    chomp($extraGrubInstallArgs);
     close FILE;
-    my $grubState = GrubState->new(name => $name, version => $version, efi => $efi, devices => $devices, efiMountPoint => $efiMountPoint );
+    my $grubState = GrubState->new(name => $name, version => $version, efi => $efi, devices => $devices, efiMountPoint => $efiMountPoint, extraGrubInstallArgs => $extraGrubInstallArgs );
     return $grubState
 }
 
-sub getDeviceTargets {
-    my @devices = ();
-    foreach my $dev ($dom->findnodes('/expr/attrs/attr[@name = "devices"]/list/string/@value')) {
-        $dev = $dev->findvalue(".") or die;
-        push(@devices, $dev);
-    }
-    return @devices;
-}
-my @deviceTargets = getDeviceTargets();
+my @deviceTargets = getList('devices');
 my $prevGrubState = readGrubState();
 my @prevDeviceTargets = split/,/, $prevGrubState->devices;
+my @extraGrubInstallArgs = getList('extraGrubInstallArgs');
+my @prevExtraGrubInstallArgs = @{decode_json($prevGrubState->extraGrubInstallArgs)};
 
 my $devicesDiffer = scalar (List::Compare->new( '-u', '-a', \@deviceTargets, \@prevDeviceTargets)->get_symmetric_difference());
+my $extraGrubInstallArgsDiffer = scalar (List::Compare->new( '-u', '-a', \@extraGrubInstallArgs, \@prevExtraGrubInstallArgs)->get_symmetric_difference());
```

Comment:
This line seems to be the reason why json was used?  Is this robust though?  Will the following trigger a rebuild?

```
extraGrubInstallArgs = [ "--arg1=foo --arg2=bar" ];
prevExtraGrubInstallArgs = [ "--arg1=foo" "--arg2=bar" ];
```

Reply:
> This line seems to be the reason why json was used?

No, JSON is used because the `/boot/grub/state` file remembers all arguments/parameters of the last successful `install-grub` run (to avoid unnecessary installation of e.g. the GRUB MBR on every `nixos-rebuild` run, which takes a couple seconds and would make no-op rebuilds slow), and the serialisation format that the state file used so far is line-based, which worked so far because there was only 1 word on each line. But as soon as you want to serialise something into it that is a list or can even contain newlines, it gets very messy and likely incorrect, so I use JSON instead and we'll use that going forward.

Your question motivated me to rework the the way this works, full details in https://github.com/NixOS/nixpkgs/pull/85895#issuecomment-653309518.

Reply:
>  Is this robust though? Will the following trigger a rebuild?

Yes, because these are different (and the first line will be rejected by `grub-install`), see https://github.com/NixOS/nixpkgs/pull/85895#discussion_r449349168.